Doc/library/collections.abc.rst

```diff
@@ -197,7 +197,7 @@ Collections Abstract Base Classes -- Detailed Descriptions
 ----------------------------------------------------------
 
 
-.. class:: Container
+.. class:: Container[T_co]
```

Comment:
Last time I checked, using PEP-695 syntax broke Sphinx's rendering for the `.. class::` directive (https://github.com/python/cpython/pull/104642#discussion_r1200799012), at least with the Sphinx version we have pinned for our docs. Has that changed since?

Reply:
At least it looks fine in [preview](https://cpython-previews--123544.org.readthedocs.build/en/123544/library/collections.abc.html#collections.abc.Coroutine) - what didn't work before, exactly?

Reply:
> what didn't work before, exactly?

I can't really give a better explanation than the screenshots I shared in https://github.com/python/cpython/pull/104642#discussion_r1200799012, I don't think (click the "details" drop-downs in that comment to see them)

Reply:
According to @AA-Turner in https://github.com/python/cpython/pull/104642#discussion_r1202366559, we shouldn't start using PEP-695 syntax in our docs until the `needs_sphinx` value in our `conf.py` is bumped to a version of Sphinx that supports rendering PEP-695 class declarations. `needs_sphinx` is currently at 6.2.1: https://github.com/python/cpython/blob/34ddb64d088dd7ccc321f6103d23153256caa5d4/Doc/conf.py#L87

And Sphinx added support for PEP-695 in v7.1.0 according to [their changelog](https://www.sphinx-doc.org/en/master/changes/7.1.html#features-added)

Reply:
Ough, this is really unfortunate - we need to list generic args somehow here. These classes don't inherit from `Generic` - would it be fine to pretend they do?

Reply:
I'd prefer not to... I think this is another argument in favour of the solution I proposed in https://github.com/python/cpython/pull/123544#discussion_r1739932413 -- keep in-depth discussion of type hints in the `typing`-module docs (but in a new section rather than in the references for the deprecated symbols like `typing.Generator` etc.), and link to that in-depth discussion from the `collections.abc` docs

Reply:
I already started moving generators and coroutines explanations to such a section, but it doesn't remove the need to provide type arguments right there. I do definitely remember all such basic usage guidelines, but can go to that documentation to check whether it's `Generator[YieldT, ReturnT, SendT]` or `Generator[YieldT, SendT, ReturnT]`. I'd rather not scroll through wall of explanatory text in such case, linking to a howto is less desired for focused search - generic args are critical part of "class signature", they shouldn't be buried so deep.

Reply:
Hm, `Generic` [doesn't have](https://github.com/python/typing/issues/449) a custom metaclass any longer, so maybe it's a good compromise to just pretend these classes inherit from `Generic`?

Reply:
> I'd rather not scroll through wall of explanatory text in such case, linking to a howto is less desired for focused search - generic args are critical part of "class signature", they shouldn't be buried so deep.

Makes sense. Another consideration is what these docs will look like to people less familiar with typing syntax, though -- we don't want to make the `collections.abc` docs impenetrable for those readers. I love typing as much as you do, but it's still an "optional add-on" to the language rather than something we want to enforce for all users... the declaration for `Generator`, for example, becomes quite a lot more complicated with this PR.

I worry that making this kind of change to the class declaration in the `collections.abc` docs could be quite controversial; it should possibly be discussed on Discourse first. Could we do it in a separate PR, and keep this PR focussed on adding the new section in the `typing` docs, and linking to that section from the `collections.abc` docs?

Reply:
Honestly, I'm becoming less and less of python's typing fan since the awkward PEP695 for a similar reason: too much attention and growing mindset of "sane python code should be typed". 

However, I agree that this might add more cognitive load than desired to these declaration. The problem of displaying generic args should be solved somehow, I'll open a separate ticket and revert this change here. As a quick suggestion: would you be more likely to accept adding explicit "type arguments" notice in every `collections.abc` class block? 

Something like this:

```rst
.. class:: Mapping
           MutableMapping

   ABCs for read-only and mutable :term:`mappings <mapping>`.

   Type arguments:

   * ``Mapping[KT, VT_co](Collection[KT])``
   * ``MutableMapping[KT, VT](Mapping[KT, VT])``
```

Reply:
> As a quick suggestion: would you be more likely to accept adding explicit "type arguments" notice in every `collections.abc` class block?

yeah, that looks much more appealing to me!

Reply:
Alex is right that we need to support Sphinx 6.2 (this is checked by CI).

Reply:
> ```rst
>    Type arguments:
> 
>    * ``Mapping[KT, VT_co](Collection[KT])``
>    * ``MutableMapping[KT, VT](Mapping[KT, VT])``
> ```

Seeing this, I'm slightly less enthusiastic, as it is a lot of syntactic complexity all at once. For a reader new to typing, or even experienced with using typing, but not with type theory, this is a lot to stomach. Worse, it's unclear where the reader can learn what the mysterious symbols `VT_co` or `KT` mean / stand for. Perhaps apocryphal, but should we be expecting readers to learn about type variance before realising that ``Mapping[str, str]`` can be used?

Ideally these symbols could be linked directly to individual explanations, but if not, could we add a footnote or reference to an explanation of the standard typevar names and their implications? 

A

Reply:
> Ideally these symbols could be linked directly to individual explanations, but if not, could we add a footnote or reference to an explanation of the standard typevar names and their implications?

We're no longer making these changes in this PR, so let's open a new issue after this has landed and discuss these questions there 👍

Reply:
Ah, I didn't realise, sorry!

A

Reply:
> And Sphinx added support for [PEP-695](https://peps.python.org/695) class declarations in v7.1.0 according to [their changelog](https://www.sphinx-doc.org/en/master/changes/7.1.html#features-added)

Yes but I needed to fix a bug in LaTeX docs, so ideally it should be using Sphinx 8.1 at least (and not just 7.1.x). We didn't get a lot of feedback for this feature and that's probably why the bug went unnoticed after the feature was released (I implemented the feature after Jelle asked for it).